### ISSUE TITLE
The GP can be unreasonably overconfident

### ISSUE THREAD

**Issue by martinmodrak** [NSAPH-Software/GPCERF#71]
It is possible I didn't use the package correctly, but a relatively direct modification of the examples yields reliably wrong, overconfident results.

Here's the code I am using:

```r
library(GPCERF)
library(ggplot2)
set.seed(32236)
N <- 500
use_nn <- FALSE

sim_data <- data.frame(Y = rep(NA_real_, N), treat = rnorm(N), cf1 = rnorm(N), cf2 = rnorm(N))
sim_data$Y <- rnorm(N, mean = sin(2 * sim_data$treat) + cos(2 * sim_data$cf1))

# exposure values
q1 <- stats::quantile(sim_data$treat, 0.05)
q2 <- stats::quantile(sim_data$treat, 0.95)

true_df <- data.frame(w = seq(q1, q2, length.out = 100))
true_df$y = sin(2 * true_df$w)

n_core <- 20

m_xgboost <- function(nthread = n_core, ...) {
  SuperLearner::SL.xgboost(nthread = nthread, ...)
}

m_ranger <- function(num.threads = n_core, ...){
  SuperLearner::SL.ranger(num.threads = num.threads, ...)
}

# Estimate GPS function
gps_m <- estimate_gps(cov_mt = sim_data[,-(1:2)],
                      w_all = sim_data$treat,
                      sl_lib = c("m_xgboost", "m_ranger"),
                      dnorm_log = TRUE)


w_all <- seq(q1, q2, length.out = 100)

params_lst <- list(alpha = 10 ^ seq(-2, 1/2, length.out = 10),
                   beta = 10 ^ seq(-2, 1/2, length.out = 10),
                   g_sigma = c(0.1, 0.5, 1, 5, 10),
                   tune_app = "all")

if(use_nn) {
  params_lst$n_neighbor <- 50
  params_lst$block_size <-  1e3
  cerf_gp_obj <- estimate_cerf_nngp(sim_data,
                                    w_all,
                                    gps_m,
                                    params = params_lst,
                                    nthread = 12)
} else {
  cerf_gp_obj <- estimate_cerf_gp(sim_data,
                                  w_all,
                                  gps_m,
                                  params = params_lst,
                                  nthread = n_core)
}

plot_fun <- function(object, ...) {
    # extract data
  tmp_data <- data.frame(w_vals = object$posterior$w,
                         mean_vals = object$posterior$mean,
                         sd_vals = object$posterior$sd)

  g1 <- ggplot2::ggplot(tmp_data) +
       ggplot2::geom_ribbon(ggplot2::aes(.data$w_vals,
                                y = .data$mean_vals,
                                ymin = .data$mean_vals - 1.96 * .data$sd_vals,
                                ymax = .data$mean_vals + 1.96 * .data$sd_vals),
                                fill = "blue", alpha = 0.25) +
        ggplot2::geom_line(ggplot2::aes(.data$w_vals, .data$mean_vals),
                           color = "blue", size = 1) +
        ggplot2::theme_bw() +
        ggplot2::ggtitle("Estimated CERF (gp) with credible band (1.96sd)") +
        ggplot2::xlab("Exposure level") +
        ggplot2::ylab("Population average counterfactual outcome")

  return(g1)
}

plot_fun(cerf_gp_obj) +
  geom_line(data = true_df,
            aes(x = w, y = y, color = "True"), size = 1.5)

```

Which results in:

![image](https://github.com/NSAPH-Software/GPCERF/assets/9483603/9825800a-a987-4e21-a118-bd50a41fa207)

I.e. the qualitative trend is there, but most of the true response curve is outside the uncertainty interval.

The situation does not improve when using more data and the nnGP - replace the config lines with 

```r
N <- 5000
use_nn <- TRUE
```

and you get:

![image](https://github.com/NSAPH-Software/GPCERF/assets/9483603/b7c173bf-e037-428c-9890-5ee790218a81)

Am I using the package wrong? It seems to me that the hyperparameter grid contains parameters that should fit the data well... Or is there a reason to expect poor performance for this type of relationship?

**Comment by martinmodrak** [comment]
An interesting variant of the problem arises with:

```r
N <- 500
use_nn <- FALSE

sim_data <- data.frame(Y = rep(NA_real_, N), treat = rnorm(N), cf1 = rnorm(N), cf2 = rnorm(N))
sim_data$Y <- rnorm(N, mean = cos(2 * sim_data$cf1))

# exposure values
q1 <- stats::quantile(sim_data$treat, 0.05)
q2 <- stats::quantile(sim_data$treat, 0.95)

true_df <- data.frame(w = seq(q1, q2, length.out = 100))
true_df$y = 0
```

Now the result correctly concentrate on flat response curve, but the curve is shifted away from 0:

![image](https://github.com/NSAPH-Software/GPCERF/assets/9483603/a23ea896-6095-43d4-8fe6-a972b0c809a6)

**Comment by boyuren158** [comment]
We thank the reviewer for raising these two special cases where our package appears to be overconfident. These two examples motivate a more refined strategy for the specification of hyper-parameter grids, which we will elaborate in the response to another [issue](https://github.com/NSAPH-Software/GPCERF/issues/72).

We would like to first clarify that the way the reviewer calculates the true CERF is incorrect. We know that in the first example the counterfactual outcome is generated by the following model $$E[Y(w)|X_1,X_2] = \sin(2w) + \cos(2X_1).$$ The true CERF is then $$E[Y(w)] = \sin(2w) + \mathbb E(\cos(2X_1)) = \sin(2w) + 0.133.$$ The last equality is due to 
$$X_1\sim N(0,1).$$ The same calculation applies for the second example, and the true CERF there is
$$E[Y(w)] = \mathbb E(\cos(2X_1)) = 0.133.$$ A constant shift 0.133 should be added to both examples when generating the true CERFs. This resolves the lack of coverage in the second example.
```r
library(GPCERF)
library(SuperLearner)
library(ggplot2)
set.seed(32236)

N <- 500
use_nn <- FALSE

sim_data <- data.frame(Y = rep(NA_real_, N), treat = rnorm(N), cf1 = rnorm(N), cf2 = rnorm(N))
sim_data$Y <- rnorm(N, mean = cos(2 * sim_data$cf1))

# exposure values
q1 <- stats::quantile(sim_data$treat, 0.05)
q2 <- stats::quantile(sim_data$treat, 0.95)

true_df <- data.frame(w = seq(q1, q2, length.out = 100))
true_df$y = 0.133

n_core <- 12

m_xgboost <- function(nthread = n_core, ...) {
  SuperLearner::SL.xgboost(nthread = nthread, ...)
}

m_ranger <- function(num.threads = n_core, ...){
  SuperLearner::SL.ranger(num.threads = num.threads, ...)
}

# Estimate GPS function
gps_m <- GPCERF::estimate_gps(cov_mt = sim_data[,-(1:2)],
                      w_all = sim_data$treat,
                      sl_lib = c("SL.mean", "SL.glm", "SL.gam", "SL.glm.interaction",
                                 "SL.earth", "m_xgboost", "m_ranger"),
                      dnorm_log = TRUE)

w_all <- seq(q1, q2, length.out = 100)

# test = gam( Y ~ s(treat), data = sim_data)
# test_pred = predict(test, data.frame(treat = w_all), se.fit = T)
# plot(w_all, test_pred$fit, type = "l", ylim = c(-1.5,1.5))
# lines(w_all, test_pred$fit + test_pred$se.fit*1.96, lty = 2)
# lines(w_all, test_pred$fit - test_pred$se.fit*1.96, lty = 2)
# lines(w_all, true_df$y, col = "red")

params_lst <- list(alpha = 10 ^ seq(-2, 1/2, length.out = 10),
                   beta = 10 ^ seq(-2, 1/2, length.out = 10),
                   g_sigma = c(0.1, 0.5, 1, 5, 10),
                   tune_app = "all")

if(use_nn) {
  params_lst$n_neighbor <- 50
  params_lst$block_size <-  1e3
  cerf_gp_obj <- estimate_cerf_nngp(sim_data,
                                    w_all,
                                    gps_m,
                                    params = params_lst,
                                    outcome_col = "Y",
                                    treatment_col = "treat",
                                    covariates_col = c("cf1", "cf2"),
                                    nthread = 10)
} else {
  cerf_gp_obj <- estimate_cerf_gp(sim_data,
                                  w_all,
                                  gps_m,
                                  params = params_lst,
                                  outcome_col = "Y",
                                  treatment_col = "treat",
                                  covariates_col = c("cf1", "cf2"),
                                  nthread = n_core)
}

plot_fun(cerf_gp_obj) +
  geom_line(data = true_df,
            aes(x = w, y = y, color = "True"), linewidth = 1.5)
```


We are aware that this modification would not improve the coverage of the first example. We would like to address this remaining issue of overconfidence by noting that in the first example the treatment  is independent of the two covariates  and . In other words, the data in the first example come from a randomized trial, rather than an observation study, which is the main target of our approach. In this case, the true GPS function  is independent of  and the estimated GPS may be highly unstable, especially when boosting based approaches (e.g., xgboost) are included in the ensemble. Since the hyper-parameter tuning procedure relies heavily on the estimated GPS, the selected values of hyper-parameters can be unreasonable. In fact, with the hyper-parameter grid provided in the example, our approach tends to select very small , which is equivalent to setting the signal-to-noise ratio of the model low, and as a result, the estimated curve is much flatter than it should be with a confidence band that barely covers the truth.
We propose to resolve the issue in example 1, and all scenarios where  is (nearly) independent of the list of available covariates, by tweaking the hyper-parameter grid (see also response to the comment 7). Since the GPS should not provide additional information to the estimation process, we fix the hyper-parameter for estimated GPS at a large value to essentially eliminate its influence on the final estimation. We then make sure the information for inference is mainly from the observed level of exposure  by setting the candidate range of it to be small, for example, between  and . We also focus on relatively large values of , reflecting a prior belief that there is some relationship between  and the outcome . The updated grid is

**Comment by boyuren158** [comment]
We thank the reviewer for raising these two special cases where our package appears to be overconfident. The first example motivatse a more refined strategy for the specification of hyper-parameter grids, which we will elaborate in the response to another [issue](https://github.com/NSAPH-Software/GPCERF/issues/72). The second example can be addressed by using the correct way to define the true CERF in `true_df`.

We would like to first clarify that the way the reviewer calculates the true CERF is incorrect. We know that in the first example the counterfactual outcome is generated by the following model $$E[Y(w)|X_1,X_2] = \sin(2w) + \cos(2X_1).$$ The true CERF is then $$E[Y(w)] = \sin(2w) + \mathbb E(\cos(2X_1)) = \sin(2w) + 0.133.$$ The last equality is due to 
$$X_1\sim N(0,1).$$ The same calculation applies for the second example, and the true CERF there is
$$E[Y(w)] = \mathbb E(\cos(2X_1)) = 0.133.$$ A constant shift 0.133 should be added to both examples when generating the true CERFs. This resolves the lack of coverage in the second example.
```r
true_df <- data.frame(w = seq(q1, q2, length.out = 100))
true_df$y = 0.133
```
The resulting estimated CERF covers the true CERF throughout the entire range of exposure.
![example2_res](https://github.com/NSAPH-Software/GPCERF/assets/16247209/35b1a51d-97e8-4001-b0b3-3cdb7b93360f)

This modification in calculating the true CERF does not improve the coverage of our estimated CERF in the first example. We would like to address this remaining issue of overconfidence by noting that in both examples the exposure `treat` is independent of the two covariates `cf1`  and `cf2`. In other words, the data in both examples come from a randomized trial, rather than an observational study, which is the main target of our approach. In this case, the true GPS function $p(\text{treat}|\text{cf1}, \text{cf2})$ is independent of the two covariates and only a function of the exposure. The two coordinates (exposure level and the GPS score) of the GP are duplicative and the inference should rely on the exposure alone (reducing to a GP labelled by a scalar). Moreover, the estimated GPS is not likely to capture the independence between the exposure and the covariates. Rather, it may overfit the data and become highly unstable, especially when flexible methods (e.g., xgboost) are used. Since the hyper-parameter tuning procedure relies heavily on the estimated GPS, the selected values of hyper-parameters, based on inaccurate GPS estimates, can be misleading. In fact, with the hyper-parameter grid provided in the example, our approach tends to select very small $\gamma$, which is equivalent to setting a low signal-to-noise ratio for the GP, and as a result, the estimated curve is flatter than it should be and the credible band barely covers the true CERF.

We propose to resolve the issue in example 1, and all scenarios where the exposure is (nearly) independent of the list of available covariates, by 1) adding more prediction algorithms (especially less complex ones) to SuperLearner to avoid overfitting of the GPS function and 2) tweaking the hyper-parameter grid to handle the redundancy of the GPS scores (see also response to this [issue](https://github.com/NSAPH-Software/GPCERF/issues/72)). Since the GPS does not provide additional information to the estimation process in this example, we fix the hyper-parameter for estimated GPS at a large value to essentially eliminate its influence on the final estimation. We then make sure the information for inference is mainly from the observed level of exposure by setting the candidate range of it to be small, for example, between 0.1 and 1. We also focus on relatively large values of $\gamma$, reflecting a prior belief that there is some relationship between the exposure and the outcome. This prior belief can be established by examining the marginal relationship between the exposure and the outcome through visualization. The updated grid is
```r
params_lst <- list(alpha = 100,
                   beta = 10^seq(-1,0,length.out = 10),
                   g_sigma = 10^seq(0,1,length.out = 10),
                   tune_app = "all")
```

The effect of adding more algorithms to the ensemble can be seen by an improvement of the accuracy of estimated GPS scores.
```r
gps1 <- GPCERF::estimate_gps(cov_mt = sim_data[,-(1:2)],
                     w_all = sim_data$treat,
                     sl_lib = c("SL.mean", "SL.glm", "SL.gam", "SL.glm.interaction",
                                "SL.earth", "m_xgboost", "m_ranger"),
                     dnorm_log = TRUE)
gps2 <- GPCERF::estimate_gps(cov_mt = sim_data[,-(1:2)],
                             w_all = sim_data$treat,
                             sl_lib = c("m_xgboost", "m_ranger"),
                             dnorm_log = TRUE)

gps_tru <- dnorm(sim_data$treat, mean = 0, sd = 1, log = T)
mean((gps_tru - gps1$gps$gps)^2)
mean((gps_tru - gps2$gps$gps)^2)
```
The MSE of `gps1` is 0.99, compared to 1.33 the MSE of `gps2`. And with the new hyper-parameter grid, the coverage of the credible interval is also improved.
```r
params_lst <- list(alpha = 100,
                   beta = 10^seq(-1,0,length.out = 10),
                   g_sigma = 10^seq(0,1,length.out = 10),
                   tune_app = "all")

cerf_gp_obj <- estimate_cerf_gp(sim_data,
                                  w_all,
                                  gps_m,
                                  params = params_lst,
                                  outcome_col = "Y",
                                  treatment_col = "treat",
                                  covariates_col = c("cf1", "cf2"),
                                  nthread = n_core)
```
![example1_res](https://github.com/NSAPH-Software/GPCERF/assets/16247209/a08adc0d-f6b3-4b44-a6f8-5d6e139a4f75)

When using nnGP approximation (sample size 5000), a similar improvement is also observed.
![example1_nngp_res](https://github.com/NSAPH-Software/GPCERF/assets/16247209/9b0675cb-29f1-4aa7-a750-b63d8cb2c37f)

Additionally, we repeat the simulations for 50 times to roughly quantify the frequentist coverage rate of the estimated credible bands.
```r
get_coverage = function(cerf_gp_obj, true_df){
  ub = cerf_gp_obj$posterior$mean + cerf_gp_obj$posterior$sd*1.96
  lb = cerf_gp_obj$posterior$mean - cerf_gp_obj$posterior$sd*1.96
  ((true_df$y>lb)*(true_df$y<ub))
}

set.seed(1)
params_lst <- list(alpha = 10,
                   beta = 10 ^ seq(-1, 1, length.out = 10),
                   g_sigma = 10^seq(0,1,length.out = 10),
                   tune_app = "all")
all_res = lapply(1:50, function(rep){
  sim_data <- data.frame(Y = rep(NA_real_, N), treat = rnorm(N), cf1 = rnorm(N), cf2 = rnorm(N))
  sim_data$Y <- rnorm(N, mean = sin(2 * sim_data$treat) + cos(2 * sim_data$cf1))

  gps_m <- GPCERF::estimate_gps(cov_mt = sim_data[,-(1:2)],
                                w_all = sim_data$treat,
                                sl_lib = c("SL.mean", "SL.glm", "SL.gam", "SL.earth",
                                           "SL.glm.interaction", "m_xgboost", "m_ranger"),
                                dnorm_log = TRUE)

  new <- estimate_cerf_gp(sim_data,
                                  w_all,
                                  gps_m,
                                  params = params_lst,
                                  outcome_col = "Y",
                                  treatment_col = "treat",
                                  covariates_col = c("cf1", "cf2"),
                                  nthread = n_core)
 list(data = sim_data, res = new)
})

all_cov = sapply(all_res, function(x){
  get_coverage(x$res, true_df)
})

mean(all_cov)
```
We achieve a reasonable pointwise coverage rate of 0.89. Since the credible band is defined with respect to the posterior distribution but not the sampling distribution of the data, we do not expect the coverage rate to be at least 0.95.
